### DISCUSSION TITLE
CloudFlare, Raspberry Pi 4 e processore ARMv7

### DISCUSSION THREAD

**Discussion by cldparisi** [indomus/forum#2707] [HUB: Home Assistant]
Seguendo la guida per installare CloudFlare in un docker su Rspberry (https://indomus.it/guide/collegarsi-da-remoto-a-home-assistant-via-reti-provider-tradizionali-o-che-usino-nat-via-cloudflare/#coredocker) mi sono imbattuto in un problema, dopo aver configurato il mio compose.yaml;

```
  cloudflared:
    container_name: cloudflared
    hostname: cloudflared
    image: cloudflare/cloudflared:latest
    restart: always
    volumes:
    - /home/pi/cloudflared:/home/nonroot/.cloudflared
    command: tunnel run UUID
```

quando lancio da linea di comando;

`docker compose run --rm cloudflared tunnel login`

mi torna un errore perchè non vien trovata l'immagine per il processore ARMv7:

`no matching manifest for linux/arm/v7 in the manifest list entries`

Se non sbaglio il Pi 4 (ma anche il PI 3) hanno processore ARMv8 e quindi l'errore mi spiazza un po' però se lancio alcuni comandi mi tornano dati che indicano che sia un ARMv7:

```
pi@raspberrypi:~/ha-compose $ uname -m
armv7l

pi@raspberrypi:~/ha-compose $ cat /proc/device-tree/model
Raspberry Pi 4 Model B Rev 1.2

pi@raspberrypi:~/ha-compose $ cat /proc/cpuinfo

processor       : 0
model name      : ARMv7 Processor rev 3 (v7l)
BogoMIPS        : 270.00
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xd08
CPU revision    : 3

processor       : 1
model name      : ARMv7 Processor rev 3 (v7l)
BogoMIPS        : 270.00
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xd08
CPU revision    : 3

processor       : 2
model name      : ARMv7 Processor rev 3 (v7l)
BogoMIPS        : 270.00
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xd08
CPU revision    : 3

processor       : 3
model name      : ARMv7 Processor rev 3 (v7l)
BogoMIPS        : 270.00
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xd08
CPU revision    : 3

Hardware        : BCM2711
Revision        : c03112
Serial          : 10000000d1052798
Model           : Raspberry Pi 4 Model B Rev 1.2

```

E' per caso possibile che sia un ARMv8 ma che lavori in emulazone ARMv7?

può essere dovuto al fatto che nel file di config non è presente?

`arm_64bit=1`

oppure al fatto che il Raspberry OS installato sia a 32 bit?

```
pi@raspberrypi:~/ha-compose $ getconf LONG_BIT
32
```

Ci sono soluzioni per poter installare CloudFlare?

Grazie

**Comment by snakuzzo** [comment]
Rpi3 ha armv7.
Per usare l'immagine ufficiale dovresti utilizzare il sistema operativo a 64 bit...tu hai quello a 32 bit.
Se non vuoi reinstallare tutto, per ora puoi utilizzare un'immagine non ufficiale (ma devi leggere bene la sua documentazione).
Esempio -> https://hub.docker.com/r/erisamoe/cloudflared

**Reply by cldparisi** [reply]
Temevo che fosse per il SO a 32bit messo sul mio Raspberry Pi 4.

Sono andato anche io alla ricerca di qualche immagine non ufficiale e ho trovato questa:
https://github.com/jauderho/dockerfiles/pkgs/container/cloudflared/versions?filters%5Bversion_type%5D=tagged
Sembra anche ben mantenuta, esce l'ufficiale e un paio di settimane dopo esce questa versione.

Ma oltre all'alternativa di reinstallare tutto su SO 64bit e quella di usare un'immagine non ufficiale secondo te è possibile anche installare Cloudflared direttamente sul Raspberry invece che sul Docker?
Mi sembra che di aver capito che non ci sia la compatibilità per l'immagine del docker non per il software in se
https://github.com/cloudflare/cloudflared/releases/download/2023.10.0/cloudflared-linux-arm 

Ti faccio ultima domanda, se decidessi di reinstallare tutto su SO a 64 bit, considerando che ho 5 docker (Home Assistant, Mosquitto, Eufy Security WS, Portainer, go2rtc) mi chiedo:

come faccio ad salvare tutti i dati per non dover riconfigurare tutto da zero
c'è il rischio che qualche dockr non sia compatibile con l'SO a 64 bit?

Grazie

**Reply by snakuzzo** [reply]
Abbiamo anche la guida per installarlo come servizio, non come container.

Se vuoi reinstallare tutto, basta che tieni da parte le directory montate come volumi sui singoli contIner

**Reply by cldparisi** [reply]
Alla fine ho deciso di reinstallare di mettere su un sistema a 64bit visto che probabilmente ne beneficerò anche su altre cose e nel mentre prendo un po' di dimestichezza con i Docker.

Per fare "export/import" dei docker mi dicevi che basta che mi backuppo le cartelle mappate nel mio compose e poi sul nuovo SO dopo aver installato i vari docker sostituisco le cartelle mappate con quelle salvate.

Ma li devo sostituire tutti?
Anche quelli che durante la copia a docker spento mi hanno dato violazione di accesso? Io a copiarli in qualche modo riesco ma poi ha senso sovrascriverli?

core.uuid: Error opening file /home/pi/homeassistant/.storage/core.uuid: Permission denied
onboarding: Error opening file /home/pi/homeassistant/.storage/onboarding: Permission denied
auth_provider.homeassistant: Error opening file /home/pi/homeassistant/.storage/auth_provider.homeassistant: Permission denied
core.config: Error opening file /home/pi/homeassistant/.storage/core.config: Permission denied
auth: Error opening file /home/pi/homeassistant/.storage/auth: Permission denied

C'è altro modo per fare export/import dei docker...magari usando portainer?

grazie
